### DISCUSSION TITLE
Multi Tx/Address requests

### DISCUSSION THREAD

**Discussion by nielstron** [CardanoSolutions/kupo#92] [Ideas]
### What is your idea? Provide a use case.

Kupo could support an interface to query multiple addressed at once.

For DEXs or DeFi protocols that support a variety of (parameterized, upgraded, ...) Smart Contracts, not all registered with the same stake key hash, using kupo would mean querying either the whole database or multiple addresses separately. This could be made much faster by supporting a single request with multiple addresses.

### Why is it a good idea?

Matching multiple addresses in a single query allows the database to optimize the case where a full scan of the database would be necessary. It can also improve the caching behaviour. It is probably worth testing these assumptions before implementation.

### What is the current alternative and why is it not good enough?

See above.

**Comment by KtorZ** [comment]
> Matching multiple addresses in a single query allows the database to optimize the case where a full scan of the database would be necessary.

- There's actually no queries where full scan are needed; every `/match` query is able to leverage an index. 
- The database and server supports concurrent reader connections, so it's possible to fire multiple queries in parallel from a client and get all responses nearly at the same time. 

Thus, I wonder if there's really a real use-case for this? It seems like this comes from a _supposition_ than an actual benchmark?

**Reply by nielstron** [reply]
Yes absolutely, these were some 100% unbacked claims, hence the suggestion

> It is probably worth testing these assumptions before implementation.

It might also be a good idea to simply add a small paragraph/section to the documentation regarding reading multiple addresses in parallel  - the use case is real and developers might be interested in this detail :)
